test/node/resendable-call-from-server.js

```diff
@@ -2,51 +2,74 @@
 
 const test = require('tap');
 
-const cp = require('child_process');
-const path = require('path');
-
 const jstp = require('../..');
 
-let client = createClient();
-
 const appName = 'application';
+
+let callsReceivedByClient = 0;
 const interfaces = {
   iface: {
-    method: (conn, callback) => {
-      callback(null);
+    method: (connection, callback) => {
+      callsReceivedByClient++;
+      if (callsReceivedByClient === 2) {
+        callback(null);
+      } else if (callsReceivedByClient > 2) {
+        test.fail('Server must not receive more than two call messages');
+      }
     },
   },
 };
 
-const application = new jstp.Application(appName, interfaces);
-const serverConfig = { applications: [application] };
-const server = jstp.net.createServer(serverConfig);
-
-client.on('message', ([message, ...args]) => {
-  switch (message) {
-    case 'error':
-      test.fail(args[0].message);
-      break;
-    case 'connected':
-      client.kill('SIGKILL');
-      server
-        .getClientsArray()[0]
-        .callMethodWithResend('iface', 'method', [], (error) => {
-          test.assertNot(error, 'must not return an error');
-          server.close();
-          client.kill('SIGKILL');
-          test.end();
-        });
-      client = createClient();
-      client.send(['reconnect', server.address().port, args[0]]);
-      break;
-  }
-});
-
-server.listen(0, () => {
-  client.send(['connect', server.address().port]);
-});
-
-function createClient() {
-  return cp.fork(path.join(__dirname, '../utils/resendable-call/client.js'));
+let server;
+let port;
+let connection;
+const client = {
+  session: null,
+  application: new jstp.Application(appName, interfaces),
+};
+
+function createServer(callback) {
+  const application = new jstp.Application(appName, interfaces);
+  server = jstp.net.createServer({ applications: [application] });
+  server.listen(0, () => {
+    port = server.address().port;
+    callback();
+  });
+  server.once('connect', (connection) => {
+    process.nextTick(() => {
+      connection.callMethodWithResend('iface', 'method', [], (error) => {
+        test.error(error, 'must not return an error');
+        server.close();
+        connection.close();
+      });
+    });
+  });
 }
+
+function connect(callback) {
+  jstp.net.connect(
+    appName,
+    client,
+    port,
+    (error, conn, session) => {
+      test.error(error, 'must connect without errors');
+      connection = conn;
+      client.session = session;
+      callback();
+    }
+  );
+}
+
+function reconnect() {
+  connection.close();
+  connection.once('close', () => {
+    jstp.net.reconnect(connection, port, (error, conn) => {
+      test.error(error, 'must reconnect without errors');
+      connection = conn;
+    });
+  });
+}
+
+test.plan(3);
+
+createServer(() => connect(reconnect));
```

Comment:
Maybe it's also worth deferring this to nextTick to make sure server had time to send a client a message and the client got it?